_posts/2015-12-09-docker-for-development.markdown

```diff
@@ -0,0 +1,214 @@
+
+```yaml
+gravity:
+  build: .
+  environment:
+    MEMCACHE_SERVERS: memcached
+    SOLR_URL: http://solr4:8983/solr/gravity
+    MONGO_HOST: mongodb
+    ELASTICSEARCH_URL: elasticsearch:9200
+    SMTP_PORT: 1025
+    SMTP_ADDRESS: mailcatcher
+  env_file: .env
+```
+The `.env` file is used for secrets such as Amazon Web Services credentials we don't want to put into the git repository.
+
+Our `docker-compose.yml` looks like this:
+
+
+```yaml
+mongodb:
+  image: mongo:2.4
+  command: bash -c "rm -f /data/db/mongod.lock; mongod --smallfiles --quiet --logpath=/dev/null"
+  ports: 
+  - "27017:27017"
+
+solr4:
+  image: artsy/solr4
+
+memcached:
+  image: memcached
+
+elasticsearch:
+  image: artsy/elasticsearch
+  ports:
+  - "9200:9200"
+  - "9300:9300"
+
+web:
+  extends:
+    file: common.yml
+    service: gravity
+  command: script/rails s -b 0.0.0.0 -p 80
+  ports:
+  - "80:80"
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+  - mailcatcher
+
+dj:
+  extends:
+    file: common.yml
+    service: gravity
+  command: bundle exec rake jobs:work
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+
+mailcatcher:
+  image: zolweb/docker-mailcatcher
+  ports:
+  - "1080:1080"
+```
+
+The command for the MongoDB section removes a lock file that can remain in place sometimes when the container is killed. Do not use that in production! We mount the local directory into the container with a `volumes:` command, so that local changes are reloaded in the running containers.
+
+Recently, [Ashkan Nasseri](https://github.com/ashkan18) began to move our delayed jobs from [delayed_job_mongoid](https://github.com/collectiveidea/delayed_job_mongoid) to [sidekiq](http://sidekiq.org/), which brings in Redis and another process that needs to run during development. Since we are using Docker, all we have to do is add a couple of new sections to our `docker-compose.yml` file:
+
+```yaml
+redis:
+  image: redis
+  ports:
+  - "6379:6379"
+
+sidekiq:
+  extends: 
+    file: common.yml
+    service: gravity
+  command: bundle exec sidekiq
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+  - redis
+```
+
+and add this line to `common.yml`:
+
+```yaml
+REDIS_URL: redis://redis
+```
+
+For development which involves multiple applications in separate git repositories, we use [Dusty](http://dusty.gc.com/), which was created by [GameChanger](https://gc.com/). Some of the advantages of using Dusty include the use of `rsync` so that performance is better, and a built-in nginx proxy along with modifications to your `/etc/hosts` file so that you can more easily connect to your applications. The shared volumes approach with the `volumes:` directive can be rather slow with the default Docker Toolbox setup using VirtualBox.
```

Comment:
I presume the "...can be rather slow..." comment is meant in regard to the Dusty's use of `rsync`, but the 2 are separated by the `nginx` mention so it's potentially unclear.